### PR TITLE
JAMES-2393 Integration tests for quota mailing

### PR DESCRIPTION
**Pull request by chibenwa**
Re-open of https://github.com/linagora/james-project/pull/1405

### CODE REVIEW

File: event-sourcing/event-sourcing-cassandra/pom.xml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            <artifactId>apache-james-backends-cassandra</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>apache-james-backends-cassandra</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (chibenwa) [inline event-sourcing/event-sourcing-cassandra/pom.xml]: groupid

File: event-sourcing/event-sourcing-cassandra/pom.xml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>apache-james-eventsourcing</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>apache-james-eventsourcing-cassandra</artifactId>
+
+    <name>Apache James :: Event sourcing :: Cassandra</name>
+
```
Comment (chibenwa) [inline event-sourcing/event-sourcing-cassandra/pom.xml]: description

File: event-sourcing/event-sourcing-api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>apache-james-eventsourcing</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>apache-james-eventsourcing-api</artifactId>
+
```
Comment (chibenwa) [inline event-sourcing/event-sourcing-api/pom.xml]: description

File: event-sourcing/event-sourcing-memory/pom.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>apache-james-eventsourcing</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>apache-james-eventsourcing-memory</artifactId>
+
+    <name>Apache James :: Event sourcing :: Memory</name>
+
```
Comment (chibenwa) [inline event-sourcing/event-sourcing-memory/pom.xml]: description

File: server/container/guice/event-sourcing/pom.xml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>james-server-guice</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>james-server-guice-event-sourcing</artifactId>
+
```
Comment (chibenwa) [inline server/container/guice/event-sourcing/pom.xml]: name, description

File: server/container/guice/mailbox-plugin-quota-mailing/pom.xml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>james-server-guice</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>james-server-guice-mailbox-plugin-quota-mailing</artifactId>
+
```
Comment (chibenwa) [inline server/container/guice/mailbox-plugin-quota-mailing/pom.xml]: name, description

File: event-sourcing/event-sourcing-api/pom.xml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>apache-james-eventsourcing</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>apache-james-eventsourcing-api</artifactId>
```
Comment (mbaechler) [inline event-sourcing/event-sourcing-api/pom.xml]: this part is actually "apache-james-eventsourcing-core"

File: event-sourcing/event-sourcing-api/pom.xml [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    <name>Apache James :: Event sourcing :: API</name>
```
Comment (mbaechler) [inline event-sourcing/event-sourcing-api/pom.xml]: Core

File: event-sourcing/event-sourcing-cassandra/pom.xml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    <parent>
+        <groupId>org.apache.james</groupId>
+        <artifactId>apache-james-eventsourcing</artifactId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>apache-james-eventsourcing-cassandra</artifactId>
+
+    <name>Apache James :: Event sourcing :: Cassandra</name>
+    <description>Cassandra implementation for James event sourcing system</description>
```
Comment (mbaechler) [inline event-sourcing/event-sourcing-cassandra/pom.xml]: we should have a apache-james-eventsourcing-eventstore-api project that is implemented by cassandra

**Comment by chibenwa** [comment]
test this please

File: src/site/xdoc/server/config-listeners.xml [hunk 10]
```diff
@@ -49,6 +49,8 @@
                      Please note that a <code>spamassassin.properties</code> file is needed.</li>
                  <li><code>org.apache.james.mailbox.cassandra.MailboxOperationLoggingListener</code>:
                      For Cassandra guice wiring. Provides more insights on mailbox operations</li>
+                 <li><code>org.apache.james.mailbox.quota.mailing.listeners.QuotaThresholdCrossingListener</code>:
+                     For Cassandra guice wiring. Sends emails to users exceeding 80% and 99% of there quota to warn them</li>
```
Comment (rouazana) [inline src/site/xdoc/server/config-listeners.xml]: their

**Comment by chibenwa** [comment]
```
BTW where are configured the 80 / 99 thresholds? 
```

Statically in the Guice injections. Doing it from the config is an other task.

```
Can you define different message for each threshold?
```

No. Nobody made this a requirement. Neither the PO through user stories nor other developers through grooming review. 

If it do is important to you, I can, of course, add this possibility.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
(Obiwan is offline)

**Comment by chibenwa** [comment]
Force pushed to solve conflict...

File: event-sourcing/event-sourcing-core/src/main/java/org/apache/james/eventsourcing/CommandDispatcher.java [hunk 11]
```diff
@@ -71,18 +68,12 @@ public int getRetries() {
...
-        List<? extends Event> handle(C c);
-    }
-
     private final EventBus eventBus;
     @SuppressWarnings("rawtypes")
     private final Map<Class, CommandHandler> handlers;
 
     @Inject
-    public CommandDispatcher(EventBus eventBus, Collection<CommandHandler<?>> handlers) {
+    public CommandDispatcher(EventBus eventBus, Collection<CommandHandler> handlers) {
```
Comment (rouazana) [inline event-sourcing/event-sourcing-core/src/main/java/org/apache/james/eventsourcing/CommandDispatcher.java]: I expect this change to throw a warning, no?

File: server/container/guice/event-sourcing/src/main/java/org/apache/james/modules/EventSourcingModule.java [hunk 12]
```diff
@@ -34,8 +32,6 @@
     @Override
     protected void configure() {
         bind(EventSourcingSystemImpl.class).in(Scopes.SINGLETON);
-        bind(CommandDispatcher.class).in(Scopes.SINGLETON);
-        bind(EventBus.class).in(Scopes.SINGLETON);
```
Comment (rouazana) [inline server/container/guice/event-sourcing/src/main/java/org/apache/james/modules/EventSourcingModule.java]: why is it no more needed?

File: event-sourcing/event-sourcing-core/src/main/java/org/apache/james/eventsourcing/CommandDispatcher.java (hunk 11 above)
Comment (rouazana) [reply event-sourcing/event-sourcing-core/src/main/java/org/apache/james/eventsourcing/CommandDispatcher.java]: fixed

File: event-sourcing/pom.xml [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <packaging>pom</packaging>
+    <artifactId>apache-james-eventsourcing</artifactId>
+
+    <name>Apache James :: Event sourcing</name>
+
+    <modules>
+        <module>event-sourcing-api</module>
```
Comment (mbaechler) [inline event-sourcing/pom.xml]: I'm not ok with current module split.
We should have eventsourcing-core that contains everything but the eventstore.
We should have an eventsourcing-eventstore-api module and eventsourcing-eventstore-memory / eventsourcing-eventstore-cassandra for implementations.
core would depends on eventstore-api.
Later on, we will also be able to extract the EventBus to have several implementations.

Comment (chibenwa) [reply event-sourcing/pom.xml]: Let's wait your PR

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/1418

**Pull request closed without merging**